[quant][fx] Enable output_quantized_idxs to allow multiple indexs

terrychenism opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73264

Summary:
Enable output_quantized_idxs to allow multiple indexs

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_multiple_input_quantized_output

Reviewers:

Subscribers:

Tasks:

Tags:

terrychenism committed:
[quant][fx] Enable output_quantized_idxs to allow multiple indexs

Summary:
Enable output_quantized_idxs to allow multiple indexs

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_multiple_input_quantized_output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][fx] Enable output_quantized_idxs to allow multiple indexs"


Summary:
Enable output_quantized_idxs to allow multiple indexs

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_multiple_input_quantized_output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][fx] Enable output_quantized_idxs to allow multiple indexs"


Summary:
Enable output_quantized_idxs to allow multiple indexs

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_multiple_input_quantized_output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
We should also extend the API to allow more complicated return types.  Here is an example:

```
class M(torch.nn.Module):             
    def forward(self, x):             
        x = torch.relu(x)             
        return {'x': x, 'x,x': (x, x)}
```

Instead of restricting it to indices, we need to make the API more general to handle these kinds of cases.  We don't have to have every case in v1, but we should have a format which could eventually extend to more and more cases.  Here is one example format which could work:

```
// typehint
// could support a couple of types initially, but is extendable to more types later
'output_to_dtype': Union[Any, List[Any], Tuple[Any], Dict[str, Any]]

// example of user configuring the output of the example above to different dtypes
{'x': torch.float, 'x, x': (torch.quint8, torch.float16)}
```

terrychenism commented:
> We should also extend the API to allow more complicated return types. Here is an example:
> 
> ```
> class M(torch.nn.Module):             
>     def forward(self, x):             
>         x = torch.relu(x)             
>         return {'x': x, 'x,x': (x, x)}
> ```
> 
> Instead of restricting it to indices, we need to make the API more general to handle these kinds of cases. We don't have to have every case in v1, but we should have a format which could eventually extend to more and more cases. Here is one example format which could work:
> 
> ```
> // typehint
> // could support a couple of types initially, but is extendable to more types later
> 'output_to_dtype': Union[Any, List[Any], Tuple[Any], Dict[str, Any]]
> 
> // example of user configuring the output of the example above to different dtypes
> {'x': torch.float, 'x, x': (torch.quint8, torch.float16)}
> ```

Yes, multiple dtype support will be in anther PR soon.

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -696,13 +696,6 @@ def maybe_insert_observers_before_graph_output(
     for those nodes.
     """
 
-    # TODO(future PR): update the output_quantized_idxs API to match
```
can we leave this comment in the code, since this PR is not adding support for arbitrary data structures?

terrychenism replied:
Sure

terrychenism committed:
Update on "[quant][fx] Enable output_quantized_idxs to allow multiple indexs"


Summary:
Enable output_quantized_idxs to allow multiple indexs

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_multiple_input_quantized_output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2235,6 +2235,49 @@ def test_fp32_input_fp32_output(self):
+
+    def test_quantized_input_quantized_multiple_output(self):
+        prepare_custom_config_dict = {
+            'input_quantized_idxs': [0], 'output_quantized_idxs': [0, 1]}
```
could we also test that the right thing happens if the user wants some of the indices to be in floating point?

terrychenism replied:
do you mean the case like this 'output_quantized_idxs': [0.5, 1.3] or [0.0, 1.0]?

vkuzo replied:
> do you mean the case like this 'output_quantized_idxs': [0.5, 1.3] or [0.0, 1.0]?

'output_quantized_idxs': [0]  // instead of 'output_quantized_idxs': [0, 1], meaning index 1 is in floating point

terrychenism replied:
Got it. Sure, I can add more case

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -700,8 +700,6 @@ def maybe_insert_observers_before_graph_output(
     # arbitrary data structures. There is always a single output, and
     # that output can have arbitrary nesting of values. List[int] is
     # not the right data type for this.
-    assert output_quantized_idxs == [0] or output_quantized_idxs == [], \
```
wait, no change is required to enable this? I think we need to unpack the tuple etc. to get the index?

terrychenism commented:
close it. new pr is ready https://github.com/pytorch/pytorch/pull/73816
